src/monitoring_service/service.py

```diff
@@ -147,40 +150,65 @@ def start(
                 check_gas_reserve(self.web3, self.private_key)
                 last_gas_check_block = last_confirmed_block
 
-            max_query_interval_end_block = (
-                self.context.ms_state.blockchain_state.latest_committed_block + MAX_FILTER_INTERVAL
-            )
-            # Limit the max number of blocks that is processed per iteration
-            last_block = BlockNumber(min(last_confirmed_block, max_query_interval_end_block))
-
-            self._process_new_blocks(last_block)
+            self._process_new_blocks(latest_confirmed_block=last_confirmed_block)
             self._trigger_scheduled_events()
             self._check_pending_transactions()
             self._purge_old_monitor_requests()
 
             wait_function(self.poll_interval)
 
-    def _process_new_blocks(self, last_block: BlockNumber) -> None:
-        # BCL return a new state and events related to channel lifecycle
-        new_chain_state, events = get_blockchain_events(
-            web3=self.web3,
-            contract_manager=CONTRACT_MANAGER,
-            chain_state=self.context.ms_state.blockchain_state,
-            to_block=last_block,
+    def _process_new_blocks(self, latest_confirmed_block: BlockNumber) -> None:
+        blockchain_state = self.context.ms_state.blockchain_state
+
+        # increment by one, as `latest_committed_block` has been queried last time already
+        from_block = BlockNumber(blockchain_state.latest_committed_block + 1)
+        to_block = min(
+            latest_confirmed_block,
+            # decrement by one, as `from_block` is inclusive
+            BlockNumber(from_block + blockchain_state.current_event_filter_interval - 1),
         )
 
+        try:
+            before_query = time.monotonic()
+            new_chain_state, events = get_blockchain_events(
```

Comment:
This is kinda problematic:

- the code is using the standard library deepcopy, which we know it is slow
- the code is doing a single request per address, this is not bad for the MS but to any other node that is using the same Ethereum node.
- the time check here is for the above plus the decoding of events, which I would guess is also slow

so the measurement here is not the threadshold of a single query, and it has lots of moving pieces in the middle, which makes it hard to reason about.

can you please fix these problems?

Reply:
1. The `BlockchainState` object has only tow layers, and it's only copied once for each new block. So I don't think it's urgent. It's already fixed in https://github.com/raiden-network/raiden-services/pull/785
2. This is true, but I would prefer to not touch that now but improve it later
3. It measures the whole process, doing the queries and decoding the events. I don't see why this is worse than just adapting to the actual RPC call, it's just a different piece of work we expect to be finished in the given amount of time.